### DISCUSSION TITLE
Insert an image into a drawn rectangle (square)

### DISCUSSION THREAD

**Discussion by batjka911** [ImageMagick/ImageMagick#7460] [Help] [answered]
### ImageMagick version

7,1,1-34

### Operating system, version and so on

Archlinux

### Description

I plan to make educational toys myself for my future grandchildren. In this case, talking about cubes. 

**So my plan.**
1. On a sheet of A4 draw a sweep of the cube.
2. Prepare 6 pictures.
3. Break the pictures into squares (4x3, 4x4, 5x5, etc.).
4. Insert the pictures (if necessary with rotation) into the cells of the cube sweep.
5. Print and glue the cubes together.

**Solution.**
1 Draw the sweep using "SVG Path Drawing". And we use the full path for each face of the cube. For the example I mean the standard (72 dpi). 

```
magick -size 595x842 xc:white -fill white -stroke black -strokewidth 2 -draw "path 'M 202.5,41 L 392.5,41 L 392.5,231 L 202.5,231 Z'" -fill white -stroke black -strokewidth 2 -draw "path 'M 202.5,231 L 392.5,231 L 392.5,421 L 202.5,421 Z'" -fill white -stroke black -strokewidth 2 -draw "path 'M 202.5,231 L 12.5,231 L 12.5,421 L 202.5,421 Z'" -fill white -stroke black -strokewidth 2 -draw "path 'M 392.5,231 L 582.5,231 L 582.5,421 L 392.5,421 Z'" -fill white -stroke black -strokewidth 2 -draw "path 'M 202.5,421 L 392.5,421 L 392.5,611 L 202.5,611 Z'" -fill white -stroke black -strokewidth 2 -draw "path 'M 202.5,611 L 392.5,611 L 392.5,801 L 202.5,801 Z'" aaa.png
```
(I know that you can use a file to describe the parameters for **_draw_** ).

2 and 3, Let's take the first version of the pictures (4x3) with a resolution of 1200x900. And break all 6 pictures into squares.

```
magick img1.jpg -crop 4x3@ +repage +adjoin img1_4x3_%d.png
.............
magick img6.jpg -crop 4x3@ +repage +adjoin img6_4x3_%d.png

```
4 And here is actually the question - how to insert completely fragments of pictures in the drawn cells of the cube, i.e use in the first pass all the img(1-6)_4x3_0.png 
And then we need a loop to process all 12 cubes.

**Comment by fmw42** [comment]
Please post examples of your input and steps.

**Reply by fmw42** [reply]
Is this what you want to do?  http://www.fmwconcepts.com/imagemagick/3Dbox/index.php. That is a bash shell script (free for personal use) for Imagemagick that will insert 6 images into the cube.

**Reply by batjka911** [reply]
Yes, you understood correctly that I need to insert 6 images. I tried to use **_-tile img1_4x3_0.png_** instead of **_-fill white_**. But the drawing is not inserted in full size.

**Reply by fmw42** [reply]
I am sorry, I still do not understand.  Can you show the command you are trying and explain what you expect and post the input images?

**Reply by batjka911** [reply]
From 12 cubes will be collected picture-4 cubes horizontally and 3 cubes vertically.
I divided the image into 12 parts like this. I do the same with the other 5 images (not publishing them here).
![img1](https://github.com/ImageMagick/ImageMagick/assets/34351872/ac46f820-bdf2-400b-9a49-2b7946bac90a)
```
magick img1.jpg -crop 4x3@ +repage +adjoin img1_4x3_%d.png
.............
magick img6.jpg -crop 4x3@ +repage +adjoin img6_4x3_%d.png
```
Now each 1 part of all 6 images I have to place on a cube.
![cube](https://github.com/ImageMagick/ImageMagick/assets/34351872/72decf6d-363a-42e5-afa9-3cfeb425b0b3)

 And so 12 times

**Reply by fmw42** [reply]
I don't see cubes. They are just squares.  The last image looks like you are going to do a cube map. But perhaps you want to make your cube from that image after printing and folding?

Which part are you having trouble with?  The first set of 12 images in 3 rows of 4 or the latter image the looks like a cube unfolded?

The first image can be created by using magick montage <list of images> -tile 4x3 -geometry +0+0 output_image.

The second can also be done using magick montage with -tile 3x4 by using null: for the space holders for for the empty spaces.

**Reply by batjka911** [reply]
> But perhaps you want to make your cube from that image after printing and folding?

That's exactly right.

> Which part are you having trouble with? 

> or the latter image the looks like a cube unfolded?

 I tried to use **_-tail img1_4x3_0.png_** instead of **_-fill white_**. But the **_img1_4x3_0.png_**  is not inserted in full size,but cropped.
```
magick -size 595x842 xc:white -tile img1_4x3_0.png -stroke black -strokewidth 2 -draw "path 'M 202.5,41 L 392.5,41 L 392.5,231 L 202.5,231 Z'" ................................ aaa.png
```
Full command for drawing an empty cube in step 1.

**Reply by fmw42** [reply]
Imagemagick does not scale automatically. You will have to use -resize if you you are not expecting to use your full size images?  How big are your images in pixels?  Can you post img1_4x3_0.png? Please provide the full command, do not use ............  We do not know what that means in order to reproduce your commands to test them to find out what may go wrong.  What do you think you are trying to achieve?  What are you trying to do with that command, functionally?

**Comment by snibgo** [comment]
@batjka911:
> magick img1.jpg -crop 4x3 +repage +adjoin img1_4x3_%d.png

That command reads `img1.jpg` and breaks it into images that are 4x3 pixels. Those images are not square, and are very small. I doubt if that does what you really want. (But I don't know what you do want.)

**Reply by batjka911** [reply]
Sorry - typo.
Of course **_-crop 4x3@_**.

**Comment by fmw42** [comment, marked as answer]
I am still not sure what you want. But here is using magick montage to create your unfolded cube.  I use only one image repeated, but you can replace that image with 6 different images.

Input:

![lena](https://github.com/ImageMagick/ImageMagick/assets/13458218/2f445018-9cf0-456c-8463-e8662575fe38)

Note: that montage apparently does not like to start with a null: image. Thus the result is shifted to the left.  So I pipe the result to -roll to shift it to the right one column.

`magick montage null: lena.png null: lena.png lena.png lena.png null: lena.png null: null: lena.png null: -tile 3x4 -geometry +0+0 miff:- | magick - -roll +256+0 result.png`

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/cf021348-f66f-45df-bc15-d63001312ce8)


Does this help?  Is that what you are trying to do?

**Reply by batjka911** [reply]
That works for me. Thank you.

Another question.
 -How do I determine the size of the square in millimeters instead of pixels?

**Reply by fmw42** [reply]
For many images the default density is 72 dpi (or ppi).  So multiply width or height in pixels by 25.4/72 where 25.4 is 1 inch converted to millimeters.  (1 in = 2.54 cm).  So size in millimeters = W\*25.4/72 or H\*25.4/72.

If you image has a different density in dpi, then use that value in place of 72

P.S. If you search Google you will find https://cssunitconverter.com/pixel-to-mm/#:~:text=For%20instance%2C%20if%20you%20have,25.4%20%2F%2096%20%3D%20285.75%20mm. and the link to a converter on the page.

**Reply by fmw42** [reply]
Note, you cannot specify mm directly for resize. But you can do the (fx:) computation in the resize from the dimensions in pixels where w is width and h is height.

`magick image.suffix -resize "%[fx:w*25.54/72]x%[fx:h*25.54/72]" output.suffix`

**Reply by batjka911** [reply]
Thank you for the detailed explanation.

**Comment by fmw42** [comment]
This works without the pipe to -roll if you know the image size. It gets around the first image as null: placeholder issue.

`magick montage -size 256x256 null: lena.png null: lena.png lena.png lena.png null: lena.png null: null: lena.png null: -tile 3x4 -geometry +0+0 result.png`

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/0b9b3766-3acc-4e7c-8a21-4c5e3ccb108a)

**Reply by batjka911** [reply]
And in this example, what does the size 256x256 indicate? 
After the standard **_magick_** command, it's kind of confusing.

**Reply by fmw42** [reply]
It is telling montage that the null: image has a size of 256x256, the same as the actual images
